articles/service-fabric/service-fabric-work-with-reliable-collections.md

```diff
@@ -216,7 +216,27 @@ Furthermore, service code is upgraded one upgrade domain at a time. So, during a
 > While you can modify the schema of a key, you must ensure that your key's hash code and equals algorithms are stable. If you change how either of these algorithms operate, you will not be able to look up the key within the reliable dictionary ever again.
 > .NET Strings can be used as a key but use the string itself as the key--do not use the result of String.GetHashCode as the key.
 
-Alternatively, you can perform what is typically referred to as a two upgrade. With a two-phase upgrade, you upgrade your service from V1 to V2: V2 contains the code that knows how to deal with the new schema change but this code doesn't execute. When the V2 code reads V1 data, it operates on it and writes V1 data. Then, after the upgrade is complete across all upgrade domains, you can somehow signal to the running V2 instances that the upgrade is complete. (One way to signal this is to roll out a configuration upgrade; this is what makes this a two-phase upgrade.) Now, the V2 instances can read V1 data, convert it to V2 data, operate on it, and write it out as V2 data. When other instances read V2 data, they do not need to convert it, they just operate on it, and write out V2 data.
+Alternatively, you can perform a multi-phase upgrade. 
+1. Deploy a new service version that
+    - has both the original version, V1 of the data contracts, and the new version, V2 of the data contracts.
+    - During this phase, the service must continue accessing data in the original, V1 collection using V1 data contracts.
+2. Deploy a new service version where
+    - a new, V2 collection iscreated;
+    - all add, update and delete operations are performed on both V1 and V2 collections in a single transaction;
+    - all read operations are performed on the V1 collection to maintain compatibility with replicas still running the older version
+3. Copy all data from all data from the V1 collection to the V2 collection.
+    - this can be done by the same service version deployed in step 3;
+4. Deploy a new service version where

```

Comment:
Based on my understanding of the previous writeup, in addition to deploy a new service version, this could also be accomplished by "configuration upgrade" mentioned in the previous writeup. I may want to consider how we phrase it to not make the customer think they have to do another service upgrade.

Reply:
I think that would require an explanation that the service will need to define a configuration setting and handle it in code, etc. SF definition of service is code + configuration, so I think "service" covers both at a higher level. If we need to go into more details, I think we'd need a separate article or a full-blown sample.